core/ast_node.js

```diff
@@ -0,0 +1,657 @@
+ */
+Blockly.ASTNode.prototype.findOutLocationForBlock_ = function(block) {
+  var newLocation, newAstNode;
+
+  newLocation = block.previousConnection.targetConnection;
+  newAstNode = Blockly.ASTNode.createConnectionNode(newLocation);
+  if (!newLocation) {
+    newLocation = block.previousConnection;
+    newAstNode = Blockly.ASTNode.createStackNode(block);
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the element to the right of the current element in the ast.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ *     block, or workspace. Return null if there is no value to the right.
+ */
+Blockly.ASTNode.prototype.next = function() {
+  var newAstNode = null;
+
+  switch (this.type_) {
+    case Blockly.ASTNode.types.WORKSPACE:
+      var newX = this.wsCoordinate_.x + Blockly.ASTNode.wsMove_;
+      var newWsCoordinate = new goog.math.Coordinate(newX, this.wsCoordinate_.y);
+      newAstNode = Blockly.ASTNode.createWorkspaceNode(this.location_,
+          newWsCoordinate);
+      break;
+
+    case Blockly.ASTNode.types.STACK:
+      var nextTopBlock = this.navigateBetweenStacks_(true);
+      newAstNode = Blockly.ASTNode.createStackNode(nextTopBlock);
+      break;
+
+    case Blockly.ASTNode.types.OUTPUT:
+      newAstNode = Blockly.ASTNode.createBlockNode(this.location_.sourceBlock_);
+      break;
+
+    case Blockly.ASTNode.types.FIELD:
+      var parentInput = this.location_.getParentInput();
+      newAstNode = this.findNextForField_(this.location_, parentInput);
+      break;
+
+    case Blockly.ASTNode.types.INPUT:
+      var inputConnection = this.location_.connection;
+      newAstNode = this.findNextForInput_(inputConnection, this.location_);
+      break;
+
+    case Blockly.ASTNode.types.BLOCK:
+      var nextConnection = this.location_.nextConnection;
+      newAstNode = Blockly.ASTNode.createConnectionNode(nextConnection);
+      break;
+
+    case Blockly.ASTNode.types.PREVIOUS:
+      var output = this.location_.outputConnection;
+      if (output) {
+        newAstNode = Blockly.ASTNode.createConnectionNode(output);
+      } else {
+        var srcBlock = this.location_.sourceBlock_;
+        newAstNode = Blockly.ASTNode.createBlockNode(srcBlock);
+      }
+      break;
+
+    case Blockly.ASTNode.types.NEXT:
+      if (this.location_.targetBlock()) {
+        newAstNode = this.findTopASTConnection_(this.location_.targetBlock());
+      }
+      break;
+  }
+
+  return newAstNode;
+};
+
+/**
+ * Find the element one level below and all the way to the left of the current
+ * location.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ * workspace, or block.
+ */
+Blockly.ASTNode.prototype.in = function() {
+  if (!this.location_) {return null;}
+  var newAstNode = null;
+
+  switch (this.type_) {
+    case Blockly.ASTNode.types.WORKSPACE:
+      var firstBlock = this.location_.getTopBlocks()[0];
+      newAstNode = this.findTopASTConnection_(firstBlock);
```

Comment:
Going in from a workspace should select the stack, not the connection on the first Block.

And please add a test case for going in on a workspace with a top block with no connections.